File: 199_BinaryTreeRightSideView.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> res = new ArrayList<Integer>();
        if(root == null){
            return res;
        }
        Queue<TreeNode> q= new ArrayDeque<TreeNode>();
        q.offer(root);
        while(!q.isEmpty()){
            int size = q.size();
            for(int i = 0; i < size; i++){
                TreeNode t = q.poll();
                if(i == size - 1){
                    res.add(t.val);
                }
                if(t.left != null){
                    q.offer(t.left);
                }
                if(t.right != null){
                    q.offer(t.right);
                }
            }
        }
        return res;
    }
}

File: c++/199_BinaryTreeRightSideView.cpp
class Solution {
public:
vector<int> res;
    vector<int> rightSideView(TreeNode* root) {
        dfs(root, 0);
        return res;
    }

    void dfs(TreeNode* root, int u) {
        if (!root) return;
        if (u == res.size()) {
            res.push_back(root->val);
        }
        u += 1;
        dfs(root->right, u);
        dfs(root->left, u);
    }
};

class Solution {
public:
    vector<int> rightSideView(TreeNode* root) {
        if (!root) return {};
        queue<TreeNode*> q;
        vector<int> res;
        q.push(root);
        while (q.size()) {
            int size = q.size();
            for (int i = 0; i < size; i++) {
                auto t = q.front();
                if (t->left) q.push(t->left);
                if (t->right) q.push(t->right);
                if (i == size - 1) res.push_back(t->val);
                q.pop();
            }
        }
        return res;
    }
};